File: sdks/akta-docs/python/src/akta_docs/engine.py
from __future__ import annotations

import glob
import re
from datetime import date, datetime
from pathlib import Path
from typing import Any

import frontmatter
import pathspec
from markdown_it import MarkdownIt
from markdown_it.token import Token

from akta_docs.models import AktaConfig, Diagnostic, LintResult, LintSummary
from akta_docs.rule_severity import rule_number_option, rule_severity
from akta_docs.word_count import count_words

DOC_ID_PATTERN = re.compile(r"^[a-z0-9][a-z0-9._-]*$")
ISO_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)?$")
QUESTION_START = re.compile(r"^(how|what|when|where|why|who|which|can|should|does|is|are)\b", re.I)

DOC_TYPES = (
    "adr",
    "business_rule",
    "feature",
    "guide",
    "reference",
    "explanation",
    "research",
    "meta",
    "changelog",
    "technical",
    "other",
)


def has_frontmatter_block(text: str) -> bool:
    return text.lstrip().startswith("---")


def body_start_line(file_content: str, body: str) -> int:
    idx = file_content.find(body)
    if idx < 0:
        return 1
    return file_content[:idx].count("\n") + 1


def inline_text(t: Token) -> str:
    if not t.children:
        return (t.content or "").strip()
    parts: list[str] = []
    for c in t.children:
        if c.type == "text":
            parts.append(c.content or "")
        elif c.type == "code_inline":
            parts.append(c.content or "")
        elif c.type == "softbreak" or c.type == "hardbreak":
            parts.append(" ")
    return "".join(parts).strip()


def parse_body(body: str) -> list[Token]:
    md = MarkdownIt("commonmark")
    return md.parse(body, {})


def collect_heading_nodes(tokens: list[Token]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    i = 0
    while i < len(tokens):
        t = tokens[i]
        if t.type == "heading_open" and t.map:
            level = int(t.tag[1])
            hmap = t.map
            text = ""
            j = i + 1
            while j < len(tokens) and tokens[j].type != "heading_close":
                if tokens[j].type == "inline":
                    text = inline_text(tokens[j])
                j += 1
            out.append({"depth": level, "map": hmap, "text": text})
            i = j + 1
            continue
        i += 1
    return out


def find_quick_fence(tokens: list[Token]) -> Token | None:
    i = 0
    while i < len(tokens):
        t = tokens[i]
        if t.type == "heading_open" and t.tag == "h1":
            i += 1
            while i < len(tokens) and tokens[i].type != "heading_close":
                i += 1
            i += 1
            while i < len(tokens):
                t2 = tokens[i]
                if t2.type == "heading_open":
                    return None
                if t2.type == "fence":
                    info = (t2.info or "").strip().split()
                    if info and info[0] == "quick":
                        return t2
                    return None
                if t2.type == "paragraph_open":
                    i += 1
                    while i < len(tokens) and tokens[i].type != "paragraph_close":
                        i += 1
                    if i < len(tokens):
                        i += 1
                    continue
                if t2.type == "html_block":
                    i += 1
                    continue
                return None
            return None
        i += 1
    return None


def map_severity(eff: str) -> str | None:
    if eff == "off":
        return None
    return eff


def sort_diags(diags: list[Diagnostic]) -> list[Diagnostic]:
    return sorted(
        diags,
        key=lambda d: (d.file, d.line, d.column, d.rule_id),
    )


def push_diag(
    diags: list[Diagnostic],
    file: str,
    rule_id: str,
    severity: str | None,
    line: int,
    column: int,
    message: str,
    end_line: int | None = None,
    end_column: int | None = None,
) -> None:
    if severity is None:
        return
    diags.append(
        Diagnostic(
            file=file,
            line=line,
            column=column,
            rule_id=rule_id,
            severity=severity,  # type: ignore[arg-type]
            message=message,
            end_line=end_line,
            end_column=end_column,
        )
    )


def lint_meta(
    file: str,
    file_content: str,
    data: dict[str, Any],
    config: AktaConfig,
    diags: list[Diagnostic],
) -> None:
    r1 = rule_severity(config, "META-001")
    if not has_frontmatter_block(file_content):
        push_diag(
            diags,
            file,
            "META-001",
            map_severity(r1),
            1,
            1,
            "Missing YAML frontmatter (expected leading --- block).",
        )
        return

    r2 = rule_severity(config, "META-002")
    doc_id = data.get("doc_id")
    if not isinstance(doc_id, str) or not doc_id.strip():
        push_diag(
            diags,
            file,
            "META-002",
            map_severity(r2),
            2,
            1,
            "Frontmatter must include non-empty string `doc_id`.",
        )
    elif not DOC_ID_PATTERN.match(doc_id.strip()):
        push_diag(
            diags,
            file,
            "META-002",
            map_severity(r2),
            2,
            1,
            f'doc_id must match pattern {DOC_ID_PATTERN.pattern}: got "{doc_id}".',
        )

    r3 = rule_severity(config, "META-003")
    doc_type = data.get("doc_type")
    if not isinstance(doc_type, str) or not doc_type.strip():
        push_diag(
            diags,
            file,
            "META-003",
            map_severity(r3),
            2,
            1,
            f"Frontmatter must include string `doc_type` (one of: {', '.join(DOC_TYPES)}).",
        )
    elif doc_type.strip() not in DOC_TYPES:
        push_diag(
            diags,
            file,
            "META-003",
            map_severity(r3),
            2,
            1,
            f'Invalid doc_type "{doc_type}".',
        )

    r4 = rule_severity(config, "META-004")
    date_val = data.get("date")
    date_str: str | None = None
    if isinstance(date_val, datetime):
        date_str = date_val.date().isoformat()
    elif isinstance(date_val, date):
        date_str = date_val.isoformat()
    elif isinstance(date_val, str) and date_val.strip():
        date_str = date_val.strip()
    if not date_str:
        push_diag(
            diags,
            file,
            "META-004",
            map_severity(r4),
            2,
            1,
            "Frontmatter must include ISO8601 `date` (YYYY-MM-DD or full instant).",
        )
    elif not ISO_DATE.match(date_str):
        push_diag(
            diags,
            file,
            "META-004",
            map_severity(r4),
            2,
            1,
            f'date must be ISO8601: got "{date_str}".',
        )


def is_question_heading(text: str) -> bool:
    s = text.strip()
    if s.endswith("?"):
        return True
    return bool(QUESTION_START.match(s))


def lint_meta_quick(
    file: str,
    data: dict[str, Any],
    tokens: list[Token],
    config: AktaConfig,
    diags: list[Diagnostic],
    line_offset: int,
) -> None:
    sev = rule_severity(config, "META-QUICK")
    eff = map_severity(sev)
    if eff is None:
        return
    if data.get("doc_type") == "changelog":
        return
    min_w = int(rule_number_option(config, "META-QUICK", "min_words", 40))
    max_w = int(rule_number_option(config, "META-QUICK", "max_words", 80))
    quick = find_quick_fence(tokens)
    if quick is None or not quick.map:
        push_diag(
            diags,
            file,
            "META-QUICK",
            eff,
            1 + line_offset,
            1,
            "After the first H1, expected a fenced code block with language tag `quick` (```quick).",
        )
        return
    text = (quick.content or "").strip()
    words = count_words(text)
    line = quick.map[0] + 1 + line_offset
    col = 1
    if words < min_w or words > max_w:
        end_ln = quick.map[1] + line_offset
        push_diag(
            diags,
            file,
            "META-QUICK",
            eff,
            line,
            col,
            f"Quick Answer block must be {min_w}–{max_w} words; found {words}.",
            end_line=end_ln,
            end_column=None,
        )


def lint_struct008(
    file: str,
    body: str,
    headings: list[dict[str, Any]],
    data: dict[str, Any],
    config: AktaConfig,
    diags: list[Diagnostic],
    line_offset: int,
) -> None:
    sev = rule_severity(config, "STRUCT-008")
    eff = map_severity(sev)
    if eff is None:
        return
    if data.get("doc_type") == "changelog":
        return
    min_w = int(rule_number_option(config, "STRUCT-008", "min_words", 150))
    max_w = int(rule_number_option(config, "STRUCT-008", "max_words", 300))
    hs = [h for h in headings if h["depth"] in (2, 3)]
    hs.sort(key=lambda h: h["map"][0])
    body_lines = body.split("\n")
    for h in hs:
        hmap = h["map"]
        # Match TS: contentStartLine = heading.position.end.line + 1 (1-based)
        content_start_line = hmap[1] + 1
        depth = h["depth"]
        nxt = next(
            (x for x in hs if x is not h and x["map"][0] > hmap[1] and x["depth"] <= depth),
            None,
        )
        # Exclusive 0-based slice end (same as TS bodyLines.slice(a, contentEndLine))
        content_end_exclusive = nxt["map"][0] if nxt else len(body_lines)
        if content_start_line > content_end_exclusive:
            push_diag(
                diags,
                file,
                "STRUCT-008",
                eff,
                hmap[0] + 1 + line_offset,
                1,
                "Section has no body; cannot satisfy STRUCT-008 word range.",
                end_line=hmap[1] + line_offset,
            )
            continue
        slice_ = "\n".join(body_lines[content_start_line - 1 : content_end_exclusive])
        wc = count_words(slice_)
        line = hmap[0] + 1 + line_offset
        if wc < min_w or wc > max_w:
            push_diag(
                diags,
                file,
                "STRUCT-008",
                eff,
                line,
                1,
                f"H{depth} section must be {min_w}–{max_w} words; found {wc}.",
                end_line=hmap[1] + line_offset,
            )


def lint_content001(
    file: str,
    headings: list[dict[str, Any]],
    config: AktaConfig,
    diags: list[Diagnostic],
    line_offset: int,
) -> None:
    sev = rule_severity(config, "CONTENT-001")
    eff = map_severity(sev)
    if eff is None:
        return
    ratio = rule_number_option(config, "CONTENT-001", "min_question_ratio", 0.7)
    h23 = [h for h in headings if h["depth"] in (2, 3)]
    if not h23:
        return
    questions = sum(1 for h in h23 if is_question_heading(h["text"]))
    actual = questions / len(h23)
    if actual + 1e-9 < ratio:
        push_diag(
            diags,
            file,
            "CONTENT-001",
            eff,
            1 + line_offset,
            1,
            f"At least {ratio * 100:.0f}% of H2/H3 headings should be questions; "
            f"got {actual * 100:.1f}% ({questions}/{len(h23)}).",
        )


def lint_files(paths: list[str], config: AktaConfig, cwd: str) -> LintResult:
    diags: list[Diagnostic] = []
    cwd_p = Path(cwd).resolve()
    for p in paths:
        abs_p = Path(p).resolve()
        try:
            rel = str(abs_p.relative_to(cwd_p))
        except ValueError:
            rel = abs_p.name
        try:
            raw = abs_p.read_text(encoding="utf-8")
        except OSError:
            continue
        post = frontmatter.loads(raw)
        data = dict(post.metadata) if post.metadata else {}
        line_offset = body_start_line(raw, post.content) - 1
        lint_meta(rel, raw, data, config, diags)
        if not has_frontmatter_block(raw):
            continue
        body = post.content
        try:
            tokens = parse_body(body)
        except Exception:
            push_diag(
                diags,
                rel,
                "META-001",
                map_severity(rule_severity(config, "META-001")),
                1,
                1,
                "Markdown body failed to parse.",
            )
            continue
        headings = collect_heading_nodes(tokens)
        lint_meta_quick(rel, data, tokens, config, diags, line_offset)
        lint_struct008(rel, body, headings, data, config, diags, line_offset)
        lint_content001(rel, headings, config, diags, line_offset)

    diags = sort_diags(diags)
    err = sum(1 for d in diags if d.severity == "error")
    warn = sum(1 for d in diags if d.severity == "warn")
    return LintResult(
        diagnostics=diags,
        summary=LintSummary(error_count=err, warn_count=warn),
    )


def expand_lint_paths(config: AktaConfig, cwd: str) -> list[str]:
    root = Path(cwd).resolve() / config.paths.docs_root
    root = root.resolve()
    out: list[str] = []
    for g in config.paths.include_globs:
        pattern = str(root / g)
        out.extend(glob.glob(pattern, recursive=True))
    out = sorted(set(out))
    spec = pathspec.PathSpec.from_lines("gitwildmatch", config.paths.exclude_globs)
    filtered: list[str] = []
    for f in out:
        rel = str(Path(f).relative_to(root))
        if spec.match_file(rel):
            continue
        filtered.append(f)
    return sorted(filtered)

File: sdks/akta-docs/typescript/src/linter/engine.ts
import fs from "node:fs/promises";
import path from "node:path";
import matter from "gray-matter";
import { unified } from "unified";
import remarkParse from "remark-parse";
import type { Code, Heading, Root } from "mdast";
import { visit } from "unist-util-visit";
import { toString } from "mdast-util-to-string";
import fg from "fast-glob";
import type { AktaConfig } from "../config/schema.js";
import { DOC_TYPES } from "../config/schema.js";
import type { Diagnostic, LintResult, RuleId, Severity } from "../types.js";
import { countWords } from "./wordCount.js";
import { ruleNumberOption, ruleSeverity } from "./ruleSeverity.js";

const DOC_ID_PATTERN = /^[a-z0-9][a-z0-9._-]*$/;
const ISO_DATE =
  /^\d{4}-\d{2}-\d{2}(?:T\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)?$/;

const QUESTION_START =
  /^(how|what|when|where|why|who|which|can|should|does|is|are)\b/i;

function isQuestionHeading(text: string): boolean {
  const s = text.trim();
  if (s.endsWith("?")) return true;
  return QUESTION_START.test(s);
}

function mapSeverity(eff: "error" | "warn" | "off"): Severity | null {
  if (eff === "off") return null;
  return eff;
}

function sortDiagnostics(a: Diagnostic, b: Diagnostic): number {
  if (a.file !== b.file) return a.file.localeCompare(b.file);
  if (a.line !== b.line) return a.line - b.line;
  if (a.column !== b.column) return a.column - b.column;
  return a.rule_id.localeCompare(b.rule_id);
}

function bodyStartLine(fileContent: string, body: string): number {
  const idx = fileContent.indexOf(body);
  if (idx < 0) return 1;
  return fileContent.slice(0, idx).split(/\r?\n/).length;
}

function parseMarkdownBody(body: string): Root {
  const processor = unified().use(remarkParse);
  return processor.parse(body) as Root;
}

function pushDiag(
  list: Diagnostic[],
  file: string,
  ruleId: RuleId,
  severity: Severity | null,
  line: number,
  column: number,
  message: string,
  endLine?: number,
  endColumn?: number,
) {
  if (!severity) return;
  list.push({
    file,
    line,
    column,
    end_line: endLine,
    end_column: endColumn,
    rule_id: ruleId,
    severity,
    message,
  });
}

function lintMeta(
  file: string,
  fileContent: string,
  fm: matter.GrayMatterFile<string>,
  config: AktaConfig,
  diags: Diagnostic[],
  _lineOffset: number,
) {
  const r1 = ruleSeverity(config, "META-001");
  if (!matter.test(fileContent)) {
    pushDiag(
      diags,
      file,
      "META-001",
      mapSeverity(r1),
      1,
      1,
      "Missing YAML frontmatter (expected leading --- block).",
    );
    return;
  }

  const data = fm.data as Record<string, unknown>;

  const r2 = ruleSeverity(config, "META-002");
  const docId = data.doc_id;
  if (typeof docId !== "string" || !docId.trim()) {
    pushDiag(
      diags,
      file,
      "META-002",
      mapSeverity(r2),
      2,
      1,
      "Frontmatter must include non-empty string `doc_id`.",
    );
  } else if (!DOC_ID_PATTERN.test(docId.trim())) {
    pushDiag(
      diags,
      file,
      "META-002",
      mapSeverity(r2),
      2,
      1,
      `doc_id must match pattern ${DOC_ID_PATTERN}: got "${docId}".`,
    );
  }

  const r3 = ruleSeverity(config, "META-003");
  const docType = data.doc_type;
  if (typeof docType !== "string" || !docType.trim()) {
    pushDiag(
      diags,
      file,
      "META-003",
      mapSeverity(r3),
      2,
      1,
      `Frontmatter must include string \`doc_type\` (one of: ${DOC_TYPES.join(", ")}).`,
    );
  } else if (!DOC_TYPES.includes(docType as (typeof DOC_TYPES)[number])) {
    pushDiag(
      diags,
      file,
      "META-003",
      mapSeverity(r3),
      2,
      1,
      `Invalid doc_type "${docType}".`,
    );
  }

  const r4 = ruleSeverity(config, "META-004");
  const dateVal = data.date;
  let dateStr: string | undefined;
  if (dateVal instanceof Date && !Number.isNaN(dateVal.getTime())) {
    dateStr = dateVal.toISOString().slice(0, 10);
  } else if (typeof dateVal === "string" && dateVal.trim()) {
    dateStr = dateVal.trim();
  }
  if (!dateStr) {
    pushDiag(
      diags,
      file,
      "META-004",
      mapSeverity(r4),
      2,
      1,
      "Frontmatter must include ISO8601 `date` (YYYY-MM-DD or full instant).",
    );
  } else if (!ISO_DATE.test(dateStr)) {
    pushDiag(
      diags,
      file,
      "META-004",
      mapSeverity(r4),
      2,
      1,
      `date must be ISO8601: got "${dateStr}".`,
    );
  }
}

function findFirstH1Index(children: Root["children"]): number {
  return children.findIndex(
    (n) => n.type === "heading" && (n as Heading).depth === 1,
  );
}

function isIgnorableBeforeQuick(n: Root["children"][number]): boolean {
  if (n.type === "paragraph") {
    return toString(n).trim() === "";
  }
  if (n.type === "html" && typeof (n as { value?: string }).value === "string") {
    return (n as { value: string }).value.trim() === "";
  }
  return false;
}

function findQuickCodeAfterH1(tree: Root): Code | undefined {
  const { children } = tree;
  const h1 = findFirstH1Index(children);
  if (h1 < 0) return undefined;
  for (let i = h1 + 1; i < children.length; i++) {
    const n = children[i];
    if (n.type === "heading") return undefined;
    if (n.type === "code" && (n as Code).lang === "quick") {
      return n as Code;
    }
    if (isIgnorableBeforeQuick(n)) continue;
    return undefined;
  }
  return undefined;
}

function collectHeadings(tree: Root): Heading[] {
  const out: Heading[] = [];
  visit(tree, "heading", (node: Heading) => {
    out.push(node);
  });
  return out;
}

function lintMetaQuick(
  file: string,
  fm: matter.GrayMatterFile<string>,
  tree: Root,
  config: AktaConfig,
  diags: Diagnostic[],
  lineOffset: number,
) {
  const sev = ruleSeverity(config, "META-QUICK");
  const eff = mapSeverity(sev);
  if (!eff) return;

  const data = fm.data as { doc_type?: string };
  if (data.doc_type === "changelog") return;

  const minW = Math.floor(
    ruleNumberOption(config, "META-QUICK", "min_words", 40),
  );
  const maxW = Math.floor(
    ruleNumberOption(config, "META-QUICK", "max_words", 80),
  );

  const quick = findQuickCodeAfterH1(tree);
  if (!quick || !quick.position) {
    pushDiag(
      diags,
      file,
      "META-QUICK",
      eff,
      1 + lineOffset,
      1,
      "After the first H1, expected a fenced code block with language tag `quick` (```quick).",
    );
    return;
  }

  const text = String(quick.value ?? "");
  const words = countWords(text);
  const line = quick.position.start.line + lineOffset;
  const col = quick.position.start.column ?? 1;
  if (words < minW || words > maxW) {
    pushDiag(
      diags,
      file,
      "META-QUICK",
      eff,
      line,
      col,
      `Quick Answer block must be ${minW}–${maxW} words; found ${words}.`,
      quick.position.end.line + lineOffset,
      quick.position.end.column,
    );
  }
}

function lintStruct008(
  file: string,
  fm: matter.GrayMatterFile<string>,
  tree: Root,
  config: AktaConfig,
  diags: Diagnostic[],
  lineOffset: number,
) {
  const sev = ruleSeverity(config, "STRUCT-008");
  const eff = mapSeverity(sev);
  if (!eff) return;

  const data = fm.data as { doc_type?: string };
  if (data.doc_type === "changelog") return;

  const minW = Math.floor(
    ruleNumberOption(config, "STRUCT-008", "min_words", 150),
  );
  const maxW = Math.floor(
    ruleNumberOption(config, "STRUCT-008", "max_words", 300),
  );

  const headings = collectHeadings(tree)
    .filter((h) => h.depth === 2 || h.depth === 3)
    .sort(
      (a, b) =>
        (a.position?.start.line ?? 0) - (b.position?.start.line ?? 0),
    );
  const bodyLines = fm.content.split(/\r?\n/);

  for (const h of headings) {
    if (!h.position) continue;
    const contentStartLine = h.position.end.line + 1;
    const next = headings.find(
      (x) =>
        x !== h &&
        x.position &&
        h.position &&
        x.position.start.line > h.position.end.line &&
        x.depth <= h.depth,
    );
    const contentEndLine = next
      ? next.position!.start.line - 1
      : bodyLines.length;

    if (contentStartLine > contentEndLine) {
      pushDiag(
        diags,
        file,
        "STRUCT-008",
        eff,
        h.position.start.line + lineOffset,
        h.position.start.column ?? 1,
        "Section has no body; cannot satisfy STRUCT-008 word range.",
        h.position.end.line + lineOffset,
        h.position.end.column,
      );
      continue;
    }

    const slice = bodyLines
      .slice(contentStartLine - 1, contentEndLine)
      .join("\n");
    const words = countWords(slice);
    const line = h.position.start.line + lineOffset;
    const col = h.position.start.column ?? 1;
    if (words < minW || words > maxW) {
      pushDiag(
        diags,
        file,
        "STRUCT-008",
        eff,
        line,
        col,
        `H${h.depth} section must be ${minW}–${maxW} words; found ${words}.`,
        h.position.end.line + lineOffset,
        h.position.end.column,
      );
    }
  }
}

function lintContent001(
  file: string,
  tree: Root,
  config: AktaConfig,
  diags: Diagnostic[],
  lineOffset: number,
) {
  const sev = ruleSeverity(config, "CONTENT-001");
  const eff = mapSeverity(sev);
  if (!eff) return;

  const ratio = ruleNumberOption(config, "CONTENT-001", "min_question_ratio", 0.7);

  const headings = collectHeadings(tree).filter(
    (h) => h.depth === 2 || h.depth === 3,
  );
  if (headings.length === 0) return;

  let questions = 0;
  for (const h of headings) {
    const t = toString(h);
    if (isQuestionHeading(t)) questions += 1;
  }

  const actual = questions / headings.length;
  if (actual + 1e-9 < ratio) {
    pushDiag(
      diags,
      file,
      "CONTENT-001",
      eff,
      1 + lineOffset,
      1,
      `At least ${(ratio * 100).toFixed(0)}% of H2/H3 headings should be questions; got ${(actual * 100).toFixed(1)}% (${questions}/${headings.length}).`,
    );
  }
}

export async function lintFiles(
  paths: string[],
  config: AktaConfig,
  cwd: string,
): Promise<LintResult> {
  const diags: Diagnostic[] = [];

  for (const p of paths) {
    const abs = path.resolve(p);
    const rel = path.relative(cwd, abs) || path.basename(abs);
    let raw: string;
    try {
      raw = await fs.readFile(abs, "utf8");
    } catch {
      continue;
    }

    const fm = matter(raw);
    const lineOffset = bodyStartLine(raw, fm.content) - 1;
    lintMeta(rel, raw, fm, config, diags, 0);
    if (!matter.test(raw)) {
      continue;
    }

    let tree: Root;
    try {
      tree = parseMarkdownBody(fm.content);
    } catch {
      pushDiag(
        diags,
        rel,
        "META-001",
        mapSeverity(ruleSeverity(config, "META-001")),
        1,
        1,
        "Markdown body failed to parse.",
      );
      continue;
    }

    lintMetaQuick(rel, fm, tree, config, diags, lineOffset);
    lintStruct008(rel, fm, tree, config, diags, lineOffset);
    lintContent001(rel, tree, config, diags, lineOffset);
  }

  diags.sort(sortDiagnostics);
  const error_count = diags.filter((d) => d.severity === "error").length;
  const warn_count = diags.filter((d) => d.severity === "warn").length;
  return { diagnostics: diags, summary: { error_count, warn_count } };
}

export async function expandLintPaths(
  config: AktaConfig,
  cwd: string,
): Promise<string[]> {
  const root = path.resolve(cwd, config.paths.docs_root);
  const entries = await fg(config.paths.include_globs, {
    cwd: root,
    absolute: true,
    onlyFiles: true,
    ignore: config.paths.exclude_globs,
  });
  return entries.sort();
}
